### DISCUSSION TITLE
What tools do you use when developing Janet?

### DISCUSSION THREAD

**Discussion by tupini07** [janet-lang/janet#987] [Polls]
Hi all! I've recently started playing around with Janet and I've been pleasantly surprised by how easy it is to express myself with it!

I'm currently using NeoVim with https://github.com/Olical/conjure/ as my _IDE_, and find that the workflow is great, although I do miss the fact that I'm not getting autocomplete for symbols I define 😢 (though this might be because of a misconfiguration on my part).

Anyway, I saw that on [Janet's homepage](https://janet-lang.org/) there's a list of supported editors and plugins, and that got me wondering: what is the most popular development environment for other folks when it comes to Janet?

**Comment by sogaiu** [comment]
# Janet Tools / Resources

What follows is a grouped write-up of some Janet tools / resources I'm aware of.

## Editor

### Neovim/Vim

The general sense I get from following the channels is that most folks use Neovim/Vim.  I think most of them use Conjure.

### Emacs

For Emacs I know there are a few users (including me).  There is more than one major mode for Janet:

* [ALSchwalm's janet-mode](https://github.com/ALSchwalm/janet-mode)
* [Ruin0x11's janet-mode](https://github.com/Ruin0x11/janet-mode)

and there are forks of both of these (presumably because the upstream maintainers are busy or some other reason).

_Update_: There is also a work-in-progress [tree-sitter-based major mode](https://github.com/sogaiu/janet-ts-mode).

Apart from the major modes, there are at least 3 modes for repl interaction:

* [ijanet-mode](https://github.com/SerialDev/ijanet-mode)
* [inf-janet](https://github.com/velkyel/inf-janet)
* [ajrepl](https://github.com/sogaiu/ajrepl)

There is also an auxilliary package ([janet-editor-elf](https://github.com/sogaiu/janet-editor-elf)) that provides some goodies.

Overall, it's not a one-stop-shop sort of situation for Emacs + Janet.

### Kakoune

I know at least one current Kakoune user (along with at least one other person who may or may not still be using Kakoune).

### VSCode

I have heard of one person who uses a fok of the VSCode extension.

### Other

I know two others who do their own thing entirely.  One of them uses and develops [freja](https://github.com/saikyun/freja) (which is written mostly in Janet).

## Formatting

* [jfmt](https://github.com/andrewchambers/jfmt) which uses [spork/fmt](https://github.com/janet-lang/spork/blob/master/spork/fmt.janet)
* [jindt](https://github.com/sogaiu/jandent#usage) which uses [jandent](https://github.com/sogaiu/jandent) (in turn based on `spork/fmt`, but tries to do [a few things differently](https://github.com/sogaiu/jandent#features))

## Linting

* `janet` executable's built-in `-k` option
* for Kakoune there is `jakoune lint` from [jakoune](https://git.sr.ht/~pepe/jakoune/)
* for Emacs there is at least [flycheck-janet](https://github.com/sogaiu/flycheck-janet) (I think @ahungry may also have an effort too)

## Testing

See [this summary](https://github.com/janet-lang/janet/discussions/846).

## Docs

* [Official docs](https://janet-lang.org/docs/index.html)
* [uvtc's Janet Notes and Examples](http://www.unexpected-vortices.com/janet/notes-and-examples/index.html)
* [Draft Zeal Docset for Janet](https://gitlab.com/sogaiu/janet-zeal-docset)

## Code

### Examples

* [Janet repository's examples directory](https://github.com/janet-lang/janet/tree/master/examples)
* [Janet repository's test directory](https://github.com/janet-lang/janet/tree/master/test)
* [boot.janet](https://github.com/janet-lang/janet/blob/master/src/boot/boot.janet)
* [JanetDocs](https://janetdocs.com/)
* [Janet Category at Rosetta Code](https://www.rosettacode.org/wiki/Category:Janet)
* [janet-eg](https://github.com/sogaiu/janet-eg) (which can use data from JanetDocs)

### Collections / Libraries

* [The official package listing](https://github.com/janet-lang/pkgs)
* [Powered by Janet](https://goto-engineering.github.io/powered-by-janet/) (though I don't know if it's active recently)
* The 3 lists of repositories of Janet code mentioned [here](https://github.com/janet-lang/janet/discussions/563#discussioncomment-271024)

### PEG Examples

* [Janet repository's tests](https://github.com/janet-lang/janet/tree/master/test)
* [some of margaret's tests](https://github.com/sogaiu/margaret/tree/master/examples)
* [a selection of samples from folks' repositories](https://github.com/sogaiu/margaret/tree/master/samples/pegs)

## Miscellaneous

### Tree-sitter

There are 2 tree-sitter grammars for Janet ([GrayJack's](https://github.com/GrayJack/tree-sitter-janet) and [sogaiu's](https://github.com/sogaiu/tree-sitter-janet-simple)).  This is relevant for at least Neovim and Emacs.  For Neovim, the [nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter) project has made a lot of general progress but AFAIK neither Janet grammar has been integrated there.  For Emacs, one of the grammars was integrated into [elisp-tree-sitter](https://github.com/emacs-tree-sitter/elisp-tree-sitter), but IIUC there is [a separate effort to integrate tree-sitter into Emacs itself](https://archive.casouri.cat/note/2021/emacs-tree-sitter/index.html) but that [effort is still in its infancy](https://lists.gnu.org/archive/html/emacs-devel/2022-05/msg00307.html) and AFAIK almost no languages are supported there ATM.

[difftastic](https://github.com/Wilfred/difftastic) has some support for Janet via one of the two grammars.

### LSP

There is at least [one effort by JohnDoneth](https://github.com/JohnDoneth/janet-language-server).  I'm not sure what the status of it is.

**Reply by tupini07** [reply]
wow thanks a lot for the quick answer and for providing so many details! This is immensely helpful to orient a newcomer to Janet as myself. I really appreciate it, and hopefully your explanation and links will be useful for other folks as well
